scripts/browserstack.py

```diff
@@ -524,19 +562,28 @@ def getFailedTestsForBuild(buildId):
 
             exitStatus = 1
             testClassesToRun = getAllTestClassNames()
-            while numRetries >= 0:
-                testResults = runTests(
-                    appUrl, testUrl, args.is_nightly, testClassesToRun
-                )
-                print("-----------------")
-                exitStatus = testResults["exitStatus"]
-                updateObservabilityWithResults(testResults["buildId"])
-                if exitStatus == 0:
-                    break
-                else:
+            testResults = runTests(
+                appUrl, testUrl, args.is_nightly, testClassesToRun
+            )
+            print("-----------------")
+            exitStatus = testResults["exitStatus"]
+            updateObservabilityWithResults(testResults["buildId"])
+            if exitStatus != 0 and numRetries > 0:
+                os.environ["BROWSERSTACK_RERUN"] = "true"
+                print(f"Getting failed tests for build {testResults['buildId']}")
+                failedTestsDictionary = getFailedTestsForBuild(testResults["buildId"])
+                while numRetries > 0:
+                    updatedFailedTestsDictionary = {}
                     numRetries -= 1
-                    testClassesToRun = getFailedTestsForBuild(testResults["buildId"])
-                    os.environ["BROWSERSTACK_RERUN"] = "true"
+                    for failedDevice in failedTestsDictionary:
+                        testResults = runTestsForFailure(appUrl, testUrl, failedDevice, failedTestsDictionary[failedDevice])
```

Comment:
I'd like to run all devices in parallel as a follow up. Right now this code runs each device in sequence.